File: ironpythoninaction/chapter15/15.2/BasicEmbeddingVB/BasicEmbeddingVB/Engine.vb
Imports System.IO
Imports System.Reflection

Imports IronPython.Hosting
Imports IronPython.Runtime
Imports Microsoft.Scripting
Imports Microsoft.Scripting.Runtime
Imports Microsoft.Scripting.Hosting
Imports Microsoft.Scripting.Hosting.Providers

Public Class Engine
    Dim _engine As ScriptEngine
    Dim _runtime As ScriptRuntime
    Dim _code As CompiledCode
    Dim _scope As ScriptScope

    Public Sub New(ByVal source As String)
        _engine = Python.CreateEngine()
        _runtime = _engine.Runtime

        AddAssemblies()
        PublishModule()
        _scope = _engine.CreateScope()
        _scope.SetVariable("__name__", "__main__")

        Dim _main As Scope
        _main = HostingHelpers.GetScope(_scope)
        _runtime.Globals.SetVariable("__main__", _main)

        Dim _script As ScriptSource
        _script = _engine.CreateScriptSourceFromString(source, SourceCodeKind.Statements)
        _code = _script.Compile()
    End Sub

    Public Function Execute() As Boolean
        Try
            _code.Execute(_scope)
            Return True
        Catch ex As Exception
            Dim eo As ExceptionOperations = _engine.GetService(Of ExceptionOperations)()
            Console.Write(eo.FormatException(ex))
            Return False
        End Try
    End Function

    Public Sub SetVariable(ByVal name As String, ByVal value As Object)
        _scope.SetVariable(name, value)
    End Sub

    Public Function TryGetVariable(ByVal name As String, ByRef result As Object) As Boolean
        Return _scope.TryGetVariable(name, result)
    End Function

    Public Sub AddAssemblies()
        Dim _assembly As Assembly
        Dim libraryPath As String

        Dim fullPath As String = Assembly.GetExecutingAssembly().Location
        Dim rootDir As String = Directory.GetParent(fullPath).FullName

        libraryPath = Path.Combine(rootDir, "ClassLibrary.dll")

        _assembly = Assembly.LoadFile(libraryPath)
        _runtime.LoadAssembly(_assembly)

        _runtime.LoadAssembly(GetType(String).Assembly)
        _runtime.LoadAssembly(GetType(Uri).Assembly)
    End Sub

    Public Sub PublishModule()
        Dim _module As Scope
        Dim inner As ScriptScope

        inner = _engine.CreateScope()
        inner.SetVariable("HelloWorld", "Some string...")
        inner.SetVariable("answer", 42)

        _module = HostingHelpers.GetScope(inner)

        _runtime.Globals.SetVariable("Example", _module)
    End Sub

End Class

File: ironpythoninaction/chapter15/15.2/BasicEmbedding/BasicEmbedded/Engine.cs
using System;
using System.IO;
using System.Reflection;

using IronPython.Hosting;
using IronPython.Runtime;
using Microsoft.Scripting;
using Microsoft.Scripting.Runtime;
using Microsoft.Scripting.Hosting;
using Microsoft.Scripting.Hosting.Providers;

namespace BasicEmbedding
{
    public class Engine
    {
        ScriptEngine _engine;
        ScriptRuntime _runtime;
        CompiledCode _code;
        ScriptScope _scope;

        public Engine(string source)
        {
            _engine = Python.CreateEngine();
            _runtime = _engine.Runtime;

            AddAssemblies();
            PublishModule();
            _scope = _engine.CreateScope();
            _scope.SetVariable("__name__", "__main__");

            Scope _main = HostingHelpers.GetScope(_scope);
            _runtime.Globals.SetVariable("__main__", _main);

            ScriptSource _script = _engine.CreateScriptSourceFromString(source, SourceCodeKind.Statements);
            _code = _script.Compile();
        }

        public bool Execute()
        {
            try
            {
                _code.Execute(_scope);
                return true;
            }
            catch (Exception e)
            {
                ExceptionOperations es = _engine.GetService<ExceptionOperations>();
                Console.Write(es.FormatException(e));
                return false;
            }

        }

        public void SetVariable(string name, object value)
        {
            _scope.SetVariable(name, value);
        }

        public bool TryGetVariable(string name, out object result)
        {
            return _scope.TryGetVariable(name, out result);
        }

        public void AddAssemblies()
        {
            string fullPath = Assembly.GetExecutingAssembly().Location;
            string rootDir = Directory.GetParent(fullPath).FullName;
            string libraryPath = Path.Combine(rootDir, "ClassLibrary.dll");

            Assembly assembly = Assembly.LoadFile(libraryPath);
            _runtime.LoadAssembly(assembly);

            _runtime.LoadAssembly(typeof(String).Assembly);
            _runtime.LoadAssembly(typeof(Uri).Assembly);
        }

        public void PublishModule()
        {
            ScriptScope inner = _engine.CreateScope();
            inner.SetVariable("HelloWorld", "Some string...");
            inner.SetVariable("answer", 42);

            Scope module = HostingHelpers.GetScope(inner);

            _runtime.Globals.SetVariable("Example", module);
        }

    }
}
